iperlsys.h

```diff
@@ -562,10 +562,20 @@ struct IPerlEnvInfo
 	(*PL_Env->pGetChildIO)(PL_Env, ptr)
 #endif
 
-#else	/* PERL_IMPLICIT_SYS */
+#else	/* below is ! PERL_IMPLICIT_SYS */
+#  ifdef USE_ITHREADS
+
+     /* Use the comma operator to return 0/non-zero, while avoiding putting
+      * this in an inline function */
+#    define PerlEnv_putenv(str)	(ENV_WRITE_LOCK, (putenv(str)               \
+                                                  ? (ENV_WRITE_UNLOCK, 1)   \
+                                                  : (ENV_WRITE_UNLOCK, 0)))
 
-#define PerlEnv_putenv(str)		putenv((str))
-#define PerlEnv_getenv(str)		getenv((str))
+#    define PerlEnv_getenv(str)		mortal_getenv(str)
```

Comment:
This looks incompatible to me with the idea that it can be called from malloc initialization.

Reply:
This is why reviews are good.

Reply:
Actually, it doesn't matter.  malloc.c  isn't usually compiled on Linux, so I set the Configure option to compile it, and discovered it doesn't compile with g++, needing a cast.  I don't know about other compilers.  This is not the result of my changes.  Hence it hasn't been used for some time, unless it works on other platforms/compilers.

But anyway, at malloc initialization, PL_scopestack_ix is 0, so it doesn't try to do anything fancy, so it works.  The changes I just pushed make that even clearer.

So, I'm unsure if it is better to leave it plain getenv() or convert.  I can see arguments both ways.

Reply:
It seems risky to me, but I haven't been able to find a case where I expect it to cause a problem.  The functions that call Perl_malloc() are only valid to call after perl_alloc() is called, which uses either malloc() to allocate the interpreter block, or the IMP_PERL_SYS structure pMalloc entry, none of which should depend on the interpreter.

perl_alloc() immediately sets the TLS, so I couldn't see a way for mortal_getenv() to see a NULL my_perl either.

The only other risk I could see was a deadlock since a thread might call malloc->getenv and another getenv (which calls malloc) giving different lock orders, but the malloc.c call to PerlEnv_getenv() only happens once, and only during startup, so I can't see a real way for it to happen.  Coverity will probably complain about.